### ISSUE TITLE
Bug of find element with empty attribute in v3.4

### ISSUE THREAD

**Issue by hirofumii** [jquery/jquery#4435]
### Description ###
Hi.
After updating jQuery to 3.4, I encountered strange behavior on Edge and IE11.

Finding elements with empty attributes was successful in v3.3.
But, v3.4 and later will fail on Edge and IE11.

```html
<div id="test">
    <input name="">
</div>
```

```js
var $input = $("#test").find("[name='']");

console.log($input.length); // This will return 0, only on Edge and IE11 after updating to v3.4
```

### Link to test case ###
https://jsbin.com/jitatiduqi/1/edit?html,css,js,output

If this is a bug, I'm glad if a fix is ​​considered.
Thanks.

**Comment by timmywil** [comment]
We should look into the behavior change, but it's not surprising as Edge and IE often fill in undefined attributes with empty string, making them difficult to differentiate.

**Comment by dmethvin** [comment]
Probably related to #3550

**Comment by mgol** [comment]
I added an update at #3550. @hirofumii can you check if your issue still exists with https://code.jquery.com/jquery-git.js instead of your current copy of jQuery?

**Comment by timmywil** [comment]
I checked this against jquery-git in IE11 on browserstack and it's still failing.

**Comment by timmywil** [comment]
Ok, I figured out the difference. It seems it's due to [a recent optimization in Sizzle](https://github.com/jquery/sizzle/commit/71fe25c4c9a10d8827b845d398ca8fb7da163749) where we tried to avoid manipulating the DOM when it wasn't necessary. Turns out you get the right result if you add the ID of the context to the selector: https://jsbin.com/vasufuv/1/edit?html,js,console,output

**Comment by hirofumii** [comment]
Thanks, @mgol.

I tested it in my environment just in case, but unfortunately the problem was not solved.
Apparently there is another cause (surely Sizzle?).

**Comment by timmywil** [comment]
@hirofumii See my comment above. It is indeed Sizzle.

**Comment by hirofumii** [comment]
@timmywil Oh, sorry! I saw the comment and intended to add 😄, but not...

**Comment by mgol** [comment]
Sizzle PR: https://github.com/jquery/sizzle/pull/450

**Comment by mgol** [comment]
The PR linked above is not one for this issue.

**Comment by mgol** [comment]
@timmywil Oh, that's weird that an ID would help here. Given that observation, should we just revert the optimization (both in Sizzle & jQuery `master`), adding a relevant unit test? After my recent PRs, all browsers except for IE & Edge don't get the temporary ID added in most cases and, as we see, IE/Edge may need it anyway.

Also, from what I see the patch at https://github.com/jquery/sizzle/commit/71fe25c4c9a10d8827b845d398ca8fb7da163749 is not entirely correct as it skips the temporary ID for sibling selectors, pushing a selector with a leading combinator to qSA directly which crashes and goes to a slower Sizzle route.

**Comment by mgol** [comment]
I though about all this, my conclusions:
1. The workaround is worth to keep even for modern browsers as it avoids tokenizing in simple cases like `elem.find( ".clazz1.clazz2" )` which are most likely the vast majority of what happens in the wild. Plus, the impact in IE is significant and that's a slow browser.
1. Prefixing with an ID selector is one way to work around the issue but it doesn't help for global selectors which we can't prefix and which don't work in IE/Edge as well.
1. One workaround I've found is adding to the document & immediately removing an input with an empty name; suddenly those selectors then work. The problem is how to write a support test for it - such a test would require adding a temporary input with an empty ID to `documentElement` & in my testing that fixes the issue in IE, making the test true... Perhaps we could accept that we'll only fix the issue in jQuery 4.0 where we can just apply the workaround in IE in `setDocument()` as we'll most likely drop EdgeHTML support in this version.

See a PR implementing that last suggestion: https://github.com/jquery/jquery/pull/4510

Let's talk it through at the next meeting.

**Comment by mgol** [comment]
It looks like applying the workaround once fixes it even for other frames and the workaround even survives browser refreshes. Therefore, we'll just apply it in all browsers unconditionally; hopefully that will resolve the issue.

Two PRs are available:
* to jQuery `master` (for jQuery 4.0): https://github.com/jquery/jquery/pull/4510
* to Sizzle (for jQuery 3.5): https://github.com/jquery/sizzle/pull/461

**Comment by mgol** [comment]
Sizzle PR landed: https://github.com/jquery/sizzle/pull/461.

Reopening as we still need to update Sizzle for this to be fixed in jQuery.

**Comment by mgol** [comment]
The fix for jQuery 4.0 landed in 05184cc448f4ed7715ddd6a5d724e167882415f1. Re-opening as we still need to update Sizzle for jQuery 3.5.0.

**Comment by mgol** [comment]
Sizzle update merged in #4641. The issue is now fixed both on `master` & on `3.x-stable`.
